File: patient-manager-app/src/service/recommendation-service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpErrorResponse } from "@angular/common/http";
import { ApiUrlUtil } from "../util/api-url-util";
import { ArrayResponse } from "../model/array-response";
import { Recommendation } from "../model/recommendation";
import { DefaultUtil } from "../util/default-util";


@Injectable({providedIn: 'root'})
export class RecommendationService
{
  public getById(id: string, callback?: (data: Recommendation) => void,
      errorCallBack?: (error: HttpErrorResponse) => void): void
  {
    const baseUrl = `api/recommendation/${(id ?? "")}`;
    const url =  ApiUrlUtil.getUrl(baseUrl);
    const request = this.http.get<Recommendation>(url);
    request.subscribe({next: callback, error: errorCallBack});
  }

  public getByPatientId(patientId: string, page: number,
      callback?: (data: ArrayResponse<Recommendation[]>) => void,
      errorCallBack?: (error: HttpErrorResponse) => void): void
  {
    const skip = (page - 1) * 10;
    const pId = patientId ?? "";
    const baseUrl = `api/recommendation/by-patient-id/${pId}/${skip}/30`;
    const url = ApiUrlUtil.getUrl(baseUrl);
    const request = this.http.get<ArrayResponse<Recommendation[]>>(url);
    request.subscribe({next: callback, error: errorCallBack});
  }

  public getByPatientIdAndCompleted(patientId: string, completed: boolean,
      page: number, callback?: (data: ArrayResponse<Recommendation[]>) => void,
      errorCallBack?: (error: HttpErrorResponse) => void): void
  {
    const skip = (page - 1) * 10;
    const pId = patientId ??  DefaultUtil.getEmptyUUID();
    const noParamUrl = "api/recommendation/by-patient-id-and-completed";
    const baseUrl = `${noParamUrl}/${pId}/${completed}/${skip}/30`;
    const url = ApiUrlUtil.getUrl(baseUrl);
    const request = this.http.get<ArrayResponse<Recommendation[]>>(url);
    request.subscribe({next: callback, error: errorCallBack});
  }

  public updateCompleted(recommendation: Recommendation,
      callback?: (data: Recommendation) => void,
      errorCallBack?: (error: HttpErrorResponse) => void): void
  {
    const baseUrl = `api/recommendation/completed/${(recommendation.id ?? "")}`;
    const url = ApiUrlUtil.getUrl(baseUrl);
    const headers = {"content-type": "application/json"};
    const body = JSON.stringify(recommendation);
    const request = this.http.put<Recommendation>(url, body, {"headers": headers});
    request.subscribe({next: callback, error: errorCallBack});
  }

  public add(recommendation: Recommendation,
      callback?: (data: Recommendation) => void,
      errorCallBack?: (error: HttpErrorResponse) => void): void
  {
    const baseUrl = "api/recommendation";
    const url = ApiUrlUtil.getUrl(baseUrl);
    const headers = {"content-type": "application/json"};
    const body = JSON.stringify(recommendation);
    const request = this.http.post<Recommendation>(url, body, {"headers": headers});
    request.subscribe({next: callback, error: errorCallBack});
  }

  public update(recommendation: Recommendation,
      callback?: (data: Recommendation) => void,
      errorCallBack?: (error: HttpErrorResponse) => void): void
  {
    const baseUrl = `api/recommendation/${recommendation.id ?? ""}`;
    const url = ApiUrlUtil.getUrl(baseUrl);
    const headers = {"content-type": "application/json"};
    const body = JSON.stringify(recommendation);
    const request = this.http.put<Recommendation>(url, body, {"headers": headers});
    request.subscribe({next: callback, error: errorCallBack});
  }

  public remove(id: string, callback?: (data: Recommendation) => void,
      errorCallBack?: (error: HttpErrorResponse) => void): void
  {
    const baseUrl = `api/recommendation/${(id ?? "")}`;
    const url =  ApiUrlUtil.getUrl(baseUrl);
    const request = this.http.delete<Recommendation>(url);
    request.subscribe({next: callback, error: errorCallBack});
  }

  public validate(recommendation: Recommendation): string
  {
    if(!recommendation.description || !recommendation.description.trim())
      return "The field 'Description' can't be empty!";

    return "";
  }

  constructor(private http: HttpClient) {}
}

File: PatientManagerWebApi/Service/RecommendationService.cs
namespace PatientManagerWebAPI.Service;

using PatientManagerWebAPI.Model;
using PatientManagerWebAPI.Data;
using PatientManagerWebApi.Error;

public class RecommendationService : IRecommendationService
{
  public async Task<Recommendation?> GetById(Guid id)
  {
    var recommendation = await _repository.GetById(id);

    if(recommendation != null)
      return recommendation;

    throw new DataNotFoundException("Couldn't find a recommendation with the requested id!");
  }

  public async Task<IEnumerable<Recommendation>>
      GetByPatientId(Guid patientId, int skip, int amount)
  {
    if(skip < 0)
      throw new InvalidAttributeException("Parameter 'Skip' can't be negative!");
    
    return await _repository.GetByPatientId(patientId, skip, amount);
  }

  
  public async Task<IEnumerable<Recommendation>>
      GetByPatientIdAndCompleted(Guid patientId, bool completed, int skip, int amount)
  {
    if(skip < 0)
      throw new InvalidAttributeException("Parameter 'Skip' can't be negative!");

    return await _repository.GetByPatientIdAndCompleted(
        patientId, completed, skip, amount);
  }

  public async Task<Recommendation?> Add(Recommendation recommendation)
  {
    FixAttributes(recommendation);
    ValidateNew(recommendation);
    recommendation.CreatedAt = recommendation.CreatedAt?.ToUniversalTime();
    return await _repository.Insert(recommendation);
  }

  public async Task<Recommendation?> Update(Recommendation recommendation)
  {
    FixAttributes(recommendation);
    ValidateUpdate(recommendation);
    var result = await _repository.Update(recommendation);

    if(result > 0)
      return await GetById(recommendation.Id ?? Guid.Empty);

    throw new DataNotFoundException("Couldn't find the requested recommendation to update!");
  }

  public async Task<Recommendation?> Remove(Guid id)
  {
    var recommendation = await _repository.GetById(id);

    if(recommendation != null)
    {
      var result = await _repository.Delete(id);

      if(result > 0)
        return recommendation;
    }

    throw new DataNotFoundException("Couldn't find the requested recommendation to remove!");
  }

  private void FixAttributes(Recommendation recommendation)
  {
    recommendation.Description = recommendation.Description?.Trim();
  }

  private void ValidateNew(Recommendation recommendation)
  {
    if(recommendation.PatientId == null || recommendation.PatientId == Guid.Empty)
    {
      throw new InvalidAttributeException(
          "The recommendation field 'patientId' can't be empty!");
    }

    if(recommendation.Description == null ||
        recommendation.Description.Trim().Equals(""))
    {
      throw new InvalidAttributeException(
          "The recommendation field 'description' can't be empty!");
    }
  }

  private void ValidateUpdate(Recommendation recommendation)
  {
    if("".Equals(recommendation.Description?.Trim()))
    {
      throw new InvalidAttributeException(
          "The recommendation field 'description' can't be empty!");
    }
  }

  public RecommendationService(IRecommendationRepository repository)
  {
    _repository = repository;
  }


  private readonly IRecommendationRepository _repository;
}